tngo application structure (#1035)
plotly/dash-docs: dash_docs/chapters/dash_enterprise/application_structure/checks.md

## CHECKS

`CHECKS` is an optional text file that allows you to precisely modify Dash 
Enterprise's app health diagnostic checks. It must be place in your app's root
directory.

> `CHECKS` is only available on Dash Enterprise Single Server. Similar functionality 
> for Dash Enterprise Kubernetes is available via the `app.json` file. 

By default, Dash Enterprise will wait 10 seconds after starting each container 
before assuming it is healthy and continuing with the deployment. Once this happens for 
all of the containers associated with the deployed Dash app, web traffic will then 

> be directed to the new containers. Dash Enterprise will then wait an additional 60 
> seconds to give time for old containers with longer running connections 
> a chance to terminate. The checks are compared to the detected `web` process in your 

Reviewer: ```suggestion
be directed to the new containers. Dash Enterprise will then wait an additional 60 
seconds before terminating the processes in the old containers; this gives time
longer running connections some time to finish processing their request.
The checks are compared to the detected `web` process in your 
```

`Procfile`.

We recommend that you include a `CHECKS` file in your app's root directory if your 

> Dash app needs more time to boot or load data into memory so as to verify if it 
> can serve traffic. 

Reviewer: ```suggestion
Dash app takes more than the default 10 seconds to boot or load data into memory in order to prevent downtime during deployments.
```

> 

Reviewer: We're missing an introduction into what the "check" is. Perhaps a simple sentence here like "The `CHECKS` file will require Dash Enterprise to make an HTTP request(s) to your Dash application to determine that it is healthy and ready to receive traffic before finishing the deployment."

> There are three settings you can modify in your `CHECKS` file:

Reviewer: There are four settings including the URL that is visited.

> 1. `WAIT` corresponds to the allocated time before the checks are performed

Reviewer: let's be more specific about what the "check" is. Like "before Dash Enterprise visits the health check URL." or something.

2. `TIMEOUT` corresponds to the time allowed for checks to be carried out
3. `ATTEMPTS`, corresponds to the number of allowed check attempts

The default `CHECKS` file values used by Dash Enterprise are the following:

```
WAIT=15
TIMEOUT=10
ATTEMPTS=3
```

In the simple example above, Dash Enterprise will wait 15 seconds before performing 
the check and will allow up to 10 seconds for a response from the app, and attempt the 

> check 3 times before marking it as a failure. 

Reviewer: As in my comment above - What is the "check" here if we're not specifying the URL to visit? I don't think that this is accurate.

Reviewer: Unless I'm missing something and there actually is some default check that is used if the URL is omitted..

Author: :hocho: I've cut this section. Dokku docs do not hint that `CHECKS` can be used without a URL. Checks can only be skipped by disabling them wit `config plugin`.

Note that omitting or commenting out `WAIT`, `TIMEOUT` or `ATTEMPTS` from a `CHECKS` file will cause Dash Enterprise to use
default setting values. 

Other instructions must be specified in the form of a relative 
link, followed by content that Dash Enterprise should find in the response. 
The expected content can be omitted if text content is not relevant. 

```
WAIT=15
TIMEOUT=10
ATTEMPTS=3
/<your-dash-app>/_dash-layout analytics
```

In this example, Dash Enterprise will visit the URL /<your-dash-app>/_dash-layout and check that the response contains the text "analytics". In practice, replace <your-dash-app> with the name of your Dash app on Dash Enterprise. The /_dash-layout path will return the JSON-ified content of your app.layout. Replace "analytics" with text that appears in the /_dash-layout. You can visit this URL directly to inspect its contents at https://<your-dash-enterprise>/<your-dash-app>/_dash-layout.

> If "analytics" is in the response within 10 seconds (TIMEOUT), Dash Enterprise will consider the check to be passed and will begin serving traffic to this new version. If "analytics" isn't in the response after 10 (`TIMEOUT`) seconds, Dash enterprise will try two more times (`ATTEMPTS`). If the next two attempts fail, Dash Enterprise will abort the deployment and keep the previous version of the app running.

Reviewer: ```suggestion
If "analytics" is in the response within 10 seconds (`TIMEOUT`), Dash Enterprise will consider the check to be passed and will begin serving traffic to this new version. If "analytics" isn't in the response after 10 (`TIMEOUT`) seconds, Dash enterprise will try two more times (`ATTEMPTS`). If the next two attempts fail, Dash Enterprise will abort the deployment and keep the previous version of the app running.
```

> The expected content (in the previous example "analytics") can be omitted. If omitted, Dash Enterprise checks that the response has a 200 Status Code.

Reviewer: Let's do a sanity check here and make sure that this is accurate.

Another use case for `CHECKS` is to avoid downtime 110 seconds (120-10) seconds 

> of downtime if an app takes a long time to start.

Reviewer: incomplete / fragmented sentence.

```python
import dash
import dash_design_kit as ddk
import dash_core_components as dcc
import dash_html_components as html
import time

# simulate long loading time
time.sleep(120)

app = dash.Dash(__name__)
server = app.server
app.layout = ddk.App([
    html.H1('hello world')
])

if __name__ == '__main__':
    app.run_server(debug=False)
```

In this example, we are simulating an app with a two-minute loading time with
`time.sleep(120)`. 

Adding checks will prevent the end user from encountering downtime while the new version of the app is getting spun up. By including a `CHECKS` file, the new version won't accept traffic until the check passes. The check can be configured to wait 120 seconds.

```
WAIT=120
TIMEOUT=10
ATTEMPTS=3

> ```

Reviewer: Missing the URL

> You can also create your own health check by writing a custom Flask endpoint that will return
> a value of your choosing. In the following example, we include  `@app.server.route('/status')` 

Reviewer: I think it's implicit that the function can return an arbitrary value & the emphasis should be more on the custom health logic included within the function rather than the return value.
```suggestion
You can also create your own health check by writing a custom Flask endpoint. In the following example, we include  `@app.server.route('/status')` 
```

that returns `OK'. In practice, you can place whatever health checks you would like within this 
function. This might include checking the uptime for 3rd party services like databases or APIs.

```python
import dash
import dash_design_kit as ddk
import dash_core_components as dcc
import dash_html_components as html

app = dash.Dash(__name__)
server = app.server
app.layout = ddk.App([
    html.H1('hello world'),
])

# Custom status file for the deployment CHECKS
@app.server.route('/status')
def update_status():

>     return 'OK'

Reviewer: ```suggestion
    # Include custom health logic in here
    return 'OK'
```

> if __name__ == '__main__':

Reviewer: ```suggestion

if __name__ == '__main__':
```

> for Dash Enterprise Kubernetes is available via the `app.json` file. 

By default, Dash Enterprise will wait 10 seconds after starting each container 
before assuming it is up and proceeding with the deployment. Once this happens for 
all of the containers associated with the deployed Dash app, web traffic will then 
be directed to the new containers. Dash Enterprise will then wait an additional 60 
seconds to give time for old containers with longer running connections 
a chance to terminate. The checks are compared to the detected `web` process in your 
`Procfile`.

We recommend that you include a `CHECKS` file in your app's root directory if your 
Dash app needs more time to boot or load data into memory so as to verify if it 
can serve traffic. 

There are three settings you can modify in your `CHECKS` file:

1. `WAIT` corresponds to the allocated time before the checks are performed
2. `TIMEOUT` corresponds to the time allowed for checks to be carried out
3. `ATTEMPTS`, corresponds to the number of allowed check attempts

The default `CHECKS` file values used by Dash Enterprise are the following:

```
WAIT=15
TIMEOUT=10

> ATTEMPTS=3

Reviewer: This isn't accurate. Dash Enterprise doesn't try any URL. It's just a 10 second timeout.

> ```

Reviewer: This isn't accurate. Dash Enterprise doesn't try any URL by default. It just waits 10 seconds and assumes that the app is healthy.

Reviewer: I don't think this comment was resolved. Here we're saying that DE uses 3 `ATTEMPTS` by default but what is `ATTEMPTS` or `TIMEOUT` referring to? There is no URL that it is attempting or timing out on.

I might replace this section with something like

"By default, Dash Enterprise simply waits 10 seconds before assuming the app is healthy. No URLs are attempted.

In the simple example above, Dash Enterprise will wait 15 seconds before performing 
the check and will allow up to 10 seconds for a response from the app, and attempt the 
check 3 times before marking it as a failure. 

Note that omitting `CHECKS` file instructions will result in Dash Enterprise skipping

> those checks. 

Reviewer: Which checks? By saying "those" it sounds like we're referring to the checks we were just talking about above which was the "default" checks. But it doesn't make sense that Dash Enterprise would skip the default checks.

```
WAIT=15
# TIMEOUT=10
ATTEMPTS=3

> ```

Reviewer: Are we missing an introduction or some context about this example?

Is the `# ` a comment? Does that work?

> You may also include instructions that are specified in the format of a relative 

Reviewer: I thought that the relative link was _required_. 

If the relative link isn't included, what is the "attempt" referring to above?

Have you tested _not_ including the relative link?

link, followed by content that Dash Enterprise should find in the response. 

> The expected content can be omitted if text content is not relevant. 

Reviewer: Let's inject an example here to explain what this syntax looks like.

```
WAIT=15
TIMEOUT=10
ATTEMPTS=3
/<your-dash-app>/_dash-layout analytics
```

In this example, Dash Enterprise will visit the URL `/<your-dash-app>/_dash-layout` and check that the response contains the text `"analytics"`. In practice, replace `<your-dash-app>` with the name of your Dash app on Dash Enterprise. The `/_dash-layout` path will return the JSON-ified content of your `app.layout`. Replace `"analytics"` with text that appears in the `/_dash-layout`. You can visit this URL directly to inspect its contents at `https://<your-dash-enterprise>/<your-dash-app>/_dash-layout`.

If `"analytics"` is in the response within 10 seconds (`TIMEOUT`), Dash Enterprise will consider the check to be passed and will begin serving traffic to this new version. If `"analytics"` isn't in the response after 10 (`TIMEOUT`) seconds, Dash enterprise will try two more times (`ATTEMPTS`). If the next two attempts fail, Dash Enterprise will abort the deployment and keep the previous version of the app running.

The expected content (in the previous example `"analytics"`) can be omitted. If omitted, Dash Enterprise checks that the response has a `200` Status Code.

(e.g if you want to check whether an image can be served). The example below checks 
that `dash-logo.png` is being served to the app.

```
WAIT=15
TIMEOUT=10
ATTEMPTS=3

/<your-dash-app>/assets/images/dash-logo.png
```

Another use case for `CHECKS` might be to verify that an app that takes

> a few minutes to load pass checks.  

Reviewer: I don't think "verify" is the right word. Dash Enterprise isn't checking to ensure that the app is taking a few minutes.

Instead, the use case is to avoid downtime 110 seconds (120-10) seconds of downtime if an app takes a long time to start.

```python
import dash
import dash_design_kit as ddk
import dash_core_components as dcc
import dash_html_components as html
import time

# simulate long loading time
time.sleep(120)

app = dash.Dash(__name__)
server = app.server
app.layout = ddk.App([
    html.H1('hello world')
])

if __name__ == '__main__':
    app.run_server(debug=False)
```

In this example, we are simulating an app with a two-minute loading time with
`time.sleep(120)`. Deploying this example without a `CHECKS` file will result in 
Dash Enterprise running an app health check with its default values.
Dash Enterprise will then be serving an app that has not yet loaded and prompting a 

> `[CRITICAL] WORKER TIMEOUT` error.

Reviewer: We're conflating two different issues here.

- `[CRITICAL] WORKER TIMEOUT` is fixed with either `--preload` or `--timeout` as per the `Procfile` section.
- Adding `CHECKS` will prevent the end user from encountering downtime while the new version of the app is getting spun up. By including a checks file, the new version won't accept traffic until the check passes. The check can be configured to wait 120 seconds.

I recommend deploying this simple example and modify the check settings and the timeout settings to verify your understanding.

The inclusion of the following `CHECKS` file resolves this issue by delaying web 
traffic long enough for the app to fully load:

```
WAIT=15
TIMEOUT=130
ATTEMPTS=2
```

You can also create your own health check by writing a custom Flask endpoint that will return
a value of your choosing. In the following example, we include  `@app.server.route('/status')` 

> that returns `OK'.

Reviewer: ```suggestion
that returns `OK'. In practice, you can place whatever health checks you would like within this function. This might include checking the uptime for 3rd party services like databases or APIs.
```

```python
import dash
import dash_design_kit as ddk
import dash_core_components as dcc
import dash_html_components as html

app = dash.Dash(__name__)
server = app.server
app.layout = ddk.App([
    html.H1('hello world'),
])

# Custom status file for the deployment CHECKS
@app.server.route('/status')
def update_status():
    return 'OK'
if __name__ == '__main__':
    app.run_server(debug=False)

```

The example app's `CHECKS` file will then resemble:

```
WAIT=15
TIMEOUT=130
ATTEMPTS=2

> <your-dash-app>/status OK

Reviewer: I believe we're missing a leading `/` here.

> ## CHECKS

Reviewer: If we're going to consolidate all of the chapters into this chapter, then lets delete the content for the other chapters and add redirects to this chapter.
